### PR TITLE
feat(components,pipelines): adding AutoML components

### PR DESCRIPTION
**Pull request by Mateusz-Switala**
**Description of your changes:**
This PR introduces the AutoML components for data processing and training categories. Furthermore, in `pipelines/training` the `automl/autogluon_tabular_training_pipeline` is introduced.

### **Tabular Data Loader** (Automl Data Loader)

A data processing component that loads tabular (CSV) data from S3-compatible storage for AutoML pipelines.

- Streams CSV data from an S3 bucket in chunks (up to 1GB) to handle large files without loading the full dataset into memory
- Supports configurable sampling: `first_n_rows`, `stratified` (by label), or `random`; method can be derived from `task_type` when not set
- Uses AWS-style credentials from environment variables (e.g. from a Kubernetes secret)
- Writes the sampled dataset to an output Dataset artifact and returns a sample configuration dictionary
- Includes unit and local integration tests

### **Tabular Train Test Split**

A component that splits a single tabular (CSV) dataset into train and test sets for AutoML workflows.

- Uses scikit-learn `train_test_split` with configurable `test_size` (default 30%) and `random_state` (default 42)
- Uses stratified splitting for binary/multiclass (by label column) and random splitting for regression
- Writes train and test CSVs to output Dataset artifacts
- Returns a sample row from the test set and the split configuration for downstream components
- Includes unit and local integration tests

### **AutoGluon Models Selection**

A training component that trains multiple AutoGluon tabular models and selects the top N by performance.

- Trains models with AutoGluon’s TabularPredictor using stacking (e.g. 3 levels) and bagging (e.g. 2 folds) on provided train/test data
- Covers multiple model families (e.g. neural nets, XGBoost, LightGBM, CatBoost, linear models)
- Evaluates all models on the test set and ranks them by the task metric (e.g. accuracy, RMSE)
- Returns top N model names, eval metric, predictor path, and model config for the refit stage
- Saves the predictor under a configurable workspace path; includes unit and local tests

### **AutoGluon Models Full Refit**

A training component that refits a chosen AutoGluon model on the full training data.

- Loads a trained TabularPredictor from a given path and refits a single model (by name) on full data
- Uses the predictor’s train/validation data for refit and a separate test dataset for evaluation
- Saves the refitted model with a `_FULL` suffix and writes artifacts under `<model_name>_FULL` (predictor, metrics, notebooks)
- Writes metrics (e.g. metrics.json, feature_importance.json, confusion_matrix for classification) and an automl predictor notebook
- Supports regression, binary, and multiclass; includes unit and local tests

### **AutoGluon Leaderboard Evaluation**

An evaluation component that builds an HTML leaderboard from multiple AutoGluon model artifacts.

- Reads pre-computed metrics from each Model artifact at `model.path / display_name / metrics / metrics.json`
- Ranks models by a chosen metric (e.g. `accuracy`, `root_mean_squared_error`) and builds a styled HTML report
- Identifies the best model and returns its name
- Outputs an HTML artifact suitable for viewing in the pipeline UI; includes unit tests

### **AutoGluon Tabular Training Pipeline**

End-to-end pipeline that runs a two-stage AutoGluon tabular workflow: train on sampled data, then refit top models on the full dataset.

- **Stage 1 – Data:** Tabular Data Loader loads CSV from S3 (with sampling); Tabular Train Test Split produces train/test datasets
- **Stage 2 – Selection:** AutoGluon Models Selection trains an ensemble on the train set, evaluates on the test set, and selects the top N models
- **Stage 3 – Refit:** AutoGluon Models Full Refit refits each selected model on the full dataset in parallel (e.g. parallelism 2)
- **Stage 4 – Evaluation:** Leaderboard Evaluation aggregates metrics from all refitted models and produces an HTML leaderboard
- Uses a configurable workspace (e.g. PVC); supports regression, binary, and multiclass with configurable `top_n` and pipeline parameters

---

## Infrastructure & Developer Experience Improvements

 - Add `automl` packages to the list of packages in the `pyproject.toml` file to streamline importing from this subcategory.
 - Add `kfp-components` to `test` extra in `pyproject.toml` and `uv.lock` files - `kfp-kubernetes` is a part of `kfp` ecosystem and can be used in pipelines-components to enhance communication between pipelines and kubernetes elements. In AutoML pipeline we use `kfp.kubernetes.use_secret_as_env` to load kubernetes secret in particular component. It seems that this package was missing from pipelines-components, but if there is a better place or approach to include it when invoking pipelines, we can revert these changes.
 - Remove `pipeline.yaml` from `.gitignore` -  We believe that the `pipeline.yaml` file located in the pipeline directory, together with the corresponding Python file containing the pipeline implementation, can improve user experience and speed up the process of testing and integrating the pipeline in the user's environment.
 - `scripts/validate_examples/validate_examples.py` - only a cosmetic change to increase readability.
 - `scripts/generate_readme/writer.py` and `scripts/generate_readme/templates/CATEGORY_README.md.j2` - when generating READMEs for subcategories the resulting README does not pass markdownlint check. The proposed changes are intended to improve the generated README files when subcategories are present so that the markdownlint check passes.

**Checklist:**

### Pre-Submission Checklist

- [x] All tests and CI checks pass
- [x] Pre-commit hooks pass without errors
- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] The title for your pull request (PR) should follow our title convention.
  [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### Additional Checklist Items for New or Updated Components/Pipelines

- [x] `metadata.yaml` includes fresh `lastVerified` timestamp
- [x] All [required files](https://github.com/kubeflow/pipelines-components/blob/main/docs/CONTRIBUTING.md#required-files)
  are present and complete
- [x] OWNERS file lists appropriate maintainers
- [x] README provides clear documentation with usage examples
- [x] Component follows `snake_case` naming convention
- [x] No security vulnerabilities in dependencies
- [ ] Containerfile included if using a custom base image

### CODE REVIEW

**Review by hbelmiro — COMMENTED** [review]
@Mateusz-Switala this PR includes changes in other directories than `pipelines` and `components`. Is this intended? If so, can you update the PR description with those changes?

Also, you need to [sign off the commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits).
To avoid modifying 130 commit messages, I’d recommend squashing (they will be squashed when the PR gets merged anyway) these into a single commit and adding Co-authored-by trailers to ensure everyone who worked on this gets proper credit. You can add these at the very end of your commit message (after a blank line) like this:

```text
Co-authored-by: Alice <alice@example.com>
Co-authored-by: Bob <bob@example.com>
```

File: .gitignore [hunk 1]
```diff
@@ -263,7 +263,6 @@ credentials/
 # Kubeflow Pipelines specific
 *.tar.gz
 component.yaml
```
Comment (hbelmiro) [inline .gitignore]: If we're removing pipeline.yaml, we should remove component.yaml too.

File: pyproject.toml [hunk 2]
```diff
@@ -51,7 +52,16 @@ Issues = "https://github.com/kubeflow/pipelines-components/issues"
 packages = [
     "kfp_components",
     "kfp_components.components",
+    "kfp_components.components.training",
```
Comment (hbelmiro) [inline pyproject.toml]: This already exists in the list.

File: components/training/automl/autogluon_leaderboard_evaluation/OWNERS
Comment (hbelmiro) [inline components/training/automl/autogluon_leaderboard_evaluation/OWNERS]: These are already root approvers. You need to set here who will actually maintain this component.

File: components/training/automl/autogluon_leaderboard_evaluation/tests/test_component_unit.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
+"""Tests for the leaderboard_evaluation component."""
+
+# Assisted-by: Cursor
```
Comment (hbelmiro) [inline components/training/automl/autogluon_leaderboard_evaluation/tests/test_component_unit.py]:
```suggestion
```

File: components/training/automl/autogluon_leaderboard_evaluation/tests/test_component_unit.py [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+                mock_html = mock.MagicMock()
+                mock_html.path = tmp_path
+
+                leaderboard_evaluation.python_func(
+                    models=mock_models,
+                    eval_metric="root_mean_squared_error",
+                    html_artifact=mock_html,
+                )
+
+                mock_df.sort_values.assert_called_once_with(by="root_mean_squared_error", ascending=False)
```
Comment (hbelmiro) [inline components/training/automl/autogluon_leaderboard_evaluation/tests/test_component_unit.py]: For `root_mean_squared_error`, lower is better. Sorting descending puts the **worst** model at rank `#1`. The test encodes this wrong direction as expected behavior — it passes when it shouldn't. If someone correctly fixed the component to sort ascending for error metrics, this test would fail and block the fix.

The same assertion appears in every other test in this file (lines 74, 138, 238).

File: pipelines/training/automl/autogluon_tabular_training_pipeline/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        "evaluates all refitted models to generate a comprehensive leaderboard with performance metrics."
+    ),
+    pipeline_config=dsl.PipelineConfig(
+        workspace=dsl.WorkspaceConfig(
+            size="1Gi",  # TODO: change to recommended size
+            kubernetes=dsl.KubernetesWorkspaceConfig(
+                pvcSpecPatch={
+                    "storageClassName": "gp3-csi",  # or 'gp3', 'fast', etc.
+                    "accessModes": ["ReadWriteOnce"],
+                }
```
[the remark this answers is not in the document]
Comment (Mateusz-Switala) [reply pipelines/training/automl/autogluon_tabular_training_pipeline/pipeline.py]: The explicit specification of the storage class name has been removed, which means that the default storage class in the user's cluster will be used.

File: components/training/automl/autogluon_leaderboard_evaluation/component.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+            }
+        )
+
+    leaderboard_df = pd.DataFrame(results).sort_values(by=eval_metric, ascending=False)
+    n = len(leaderboard_df)
+    leaderboard_df.index = pd.RangeIndex(start=1, stop=n + 1, name="rank")
+
+    html_table = leaderboard_df.to_html(classes=None, border=0, escape=True)
+
+    best_model_name = leaderboard_df.iloc[0]["model"]
```
[the remark this answers is not in the document]
Comment (Mateusz-Switala) [reply components/training/automl/autogluon_leaderboard_evaluation/component.py]: AutoGluon follows the convention higher is better and some metrics that uses the reverse strategy have "flipped" sign. To sum up, the leaderboard is correctly sorted.
